### PR TITLE
fix: fix various bugs with "Tap" events and "closeOnClick" options #11423

### LINKED ISSUE #11423: hoverable dropdown-pane won't close on an iPad
## What should happen?
Dropdown (dropdown-pane) should close when clicking outside the dropdown-pane with data-hover="true" data-hover-pane="true" data-close-on-click="true"

## What happens instead?
On iPad 11.4.1 (Safari/Chrome) the dropdown pane does not close on clicking outside the dropdown-pane.

## Test Case and/or Steps to Reproduce

Test Case:
https://codepen.io/KraftPixel/pen/OwOWxO
https://foundation.zurb.com/sites/docs/dropdown.html
Check the Hoverable Dropdown

## Context

Trying to build a mega menu that lists more than just a list of links. Works well on desktop. Failed on iPad's landscape mode.

## Your Environment

- Foundation version(s) used: Latest
- Browser(s) name and version(s): 
Chrome 68 on iOS 11.4
Safari 11 on iOS 11.4
- Operating System and version (desktop or mobile): Mobile - iPad - iOS 11.4.1

## Checklist (all required):

- [x] I have read and follow the CONTRIBUTING.md document.
- [x] There are no other issues similar to this one.
- [x] The issue title is descriptive.
- [x] The template is correctly filled.

### PR DESCRIPTION
**Pull request by ncoden**
## Description

Fix various buges related to the touch `tap` event and the `closeOnClick` options of Dropdown and Reveal.

- Closes #11423

### Fixes
* Fix the "tap" event triggering in Touch 
* Support mobile "tap" for the `closeOnClick` options in Dropdown and Reveal (#11423)
* Keep the original target (and others event datas) in `tap`/`swipe` events in Touch

### Others changes
* Update test description for the Dropdown `closeOnClick` option
* Add visual test for the Reveal `closeOnClick` option

### 💥 BREAKING CHANGE
* `foundation.utils.touch.js` is now required for `foundation.dropdown.js` and `foundation.reveal.js`. If you import the Foundation Dropdown or Reveal plugins manually, make sure to import the "Touch" utility as well.

## Types of changes

- [ ] Documentation
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing
      functionality to change)

## Checklist (all required):

- [x] I have read and follow the [CONTRIBUTING](CONTRIBUTING.md) document.
- [x] There are no other pull request similar to this one.
- [x] The pull request title is descriptive.
- [x] The template is fully and correctly filled.
- [x] The pull request targets the right branch (`develop` or `develop-v...`).
- [x] My commits are correctly titled and contain all relevant information.
- [x] My code follows the code style of this project.
- [x] I have updated the documentation accordingly to my changes (if relevant).
- [x] I have added tests to cover my changes (if relevant).
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by ncoden** [comment]
Note: a patch for `v6.5.0` could be created with only the non-breaking fixes for the Touch utility.

File: js/foundation.util.touch.js [hunk 1]
```diff
@@ -11,25 +11,36 @@ var startPosX,
...
+    didMoved = false;
 
-function onTouchEnd() {
-  //  alert(this);
+function onTouchEnd(e) {
   this.removeEventListener('touchmove', onTouchMove);
   this.removeEventListener('touchend', onTouchEnd);
+
+
+  // If the touch did not moved, consider it as a "tap"
```
Comment (IamManchanda) [inline js/foundation.util.touch.js]: Too much spaces ... 1 space b/w code please

Comment (ncoden) [reply js/foundation.util.touch.js]: 👍

Comment (ncoden) [reply js/foundation.util.touch.js]: resolved

**Review by DanielRuf — APPROVED** [review]
LGTM

File: js/foundation.util.touch.js [hunk 2]
```diff
@@ -11,25 +11,35 @@ var startPosX,
...
+    isMoving = false,
+    didMoved = false;
 
-function onTouchEnd() {
-  //  alert(this);
+function onTouchEnd(e) {
   this.removeEventListener('touchmove', onTouchMove);
   this.removeEventListener('touchend', onTouchEnd);
+
+  // If the touch did not moved, consider it as a "tap"
```
Comment (DanielRuf) [inline js/foundation.util.touch.js]: Oh,

moved => move

File: test/visual/reveal/close-on-click.html [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  </head>
+  <body>
+
+    <div class="grid-container">
+      <div class="grid-x grid-padding-x">
+        <div class="cell">
+
+          <h1>Reveal: closeOnClick option</h1>
+
+          <p>This modal should not close when clicking/taping outside of it.</p>
```
Comment (DanielRuf) [inline test/visual/reveal/close-on-click.html]: taping => tapping

File: test/visual/reveal/close-on-click.html [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+          <p><button class="button" data-open="exampleModal1">Open modal</button></p>
+
+          <div class="reveal" id="exampleModal1" data-reveal data-close-on-click="false">
+            <p>Clicking/taping outside of me SHOULD NOT close me.</p>
```
Comment (DanielRuf) [inline test/visual/reveal/close-on-click.html]: taping => tapping

File: test/visual/reveal/close-on-click.html [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+            <p><button class="button" data-close="exampleModal1">Close</button></p>
+          </div>
+
+          <hr>
+
+          <p>This modal should close when clicking (desktop) or taping without swiping (mobile) outside of it.</p>
```
Comment (DanielRuf) [inline test/visual/reveal/close-on-click.html]: taping => tapping

File: test/visual/reveal/close-on-click.html [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+          <p><button class="button" data-open="exampleModal2">Open modal</button></p>
+
+          <div class="reveal" id="exampleModal2" data-reveal data-close-on-click="true">
+            <p>Clicking (desktop) or taping without swiping (mobile) outside of me SHOULD close me.</p>
```
Comment (DanielRuf) [inline test/visual/reveal/close-on-click.html]: taping => tapping

**Review by DanielRuf — CHANGES_REQUESTED** [review]

File: js/foundation.util.touch.js (hunk 2 above)
Comment (ncoden) [reply js/foundation.util.touch.js]: Thank you for the review. Did you tested the code ?

**Comment by ncoden** [comment]
@DanielRuf Fixed. Did you tested the code ?

**Comment by DanielRuf** [comment]
Yes, I did test the code, works with tap and click so far (emulated and on my Android).

**Comment by ncoden** [comment]
@DanielRuf Fixed.

**Comment by ncoden** [comment]
Thank you for the review. I guess you can approve it again now :)

**Review by DanielRuf — APPROVED** [review]

**Comment by ncoden** [comment]
> Note: a patch for v6.5.0 could be created with only the non-breaking fixes for the Touch utility.

See https://github.com/zurb/foundation-sites/pull/11531

**Pull request merged**